docs/goals_and_soa.md

```diff
@@ -0,0 +1,392 @@
+by employing constraint equivalence check,
+and verifying that any deviation from generated transitions
+is not possible. Another solution might be mutation testing.
+Such kind of vulnerabilities are now the most complex kind of attack
+to consider as part of DApp design or auditing.
+Thus, automaing it may be benefitial.
+
+### Time and stages handling
+
+Plutus time is described by allowed intervals.
+Such abstraction is prone to off-by-one and similar kind of errors.
+At least three errors of this kind are known in Cardano/Plutus
+interval logic, showing that this is not simple to implement
+correctly.
+
+Another problem is keeping time logic in sync between on-
+and off-chain code. This is even more hard given that Plutus time
+to slot conversion is not obvious to implement correctly.
+Slot time differences and overall need to make test cases match
+real blockchain behaviour may lead to flacky test behaviour.
+
+Our script stages abstraction cover all those kind of problems.
+
+* @todo #3: document problems with slots
+* @todo #3: bug example
+
+
+## Logic duplication and spec subtleness
+
+### Human-readable specification
+
+Designing, understanding and auditing any non-trivial DApp
+is almost impossible without human-readable spec.
+That is why in all reference DApps either used spec in development,
+or was asked to provide specification by auditor.
+Tweag and MLabs audits specificalli list validating provided specification
+as one of main tasks of audit.
+
+This leads to lot of cruft with writing, linking and updating specifications. Quite often not only one, but up to three levels of spec granularity would be benefitial for project, which worstens situation.
+We observe great amount of cruft and spec rot in all projects we were personally involved. Such kind of cruft is often present
+not only on level of single developers, but on communication
+
+Such kind of human-readable specs are very much similar to our model,
+they almost always include enumeration of possible transitions
+and/or state-machine diagrams.
+Our project will generate such kind of specifications automatically
+as Markdown file with Mermaid diagrams.
+Adding usecases scenarios generated from one specified in tests
+is much less obvious to implement,
+and out of scope of current Catalyst project,
+but it is very much possible feature as well.
+
+### On-/Off-chain and spec code duplication
+
+@todo #3: more integration point code duplication examples
+
+### Correct off-chain Tx construction logic
+
+A lot of on-chain problems, like timing and coin change issues
+have their counterpart on Tx sumbmission side.
+
+@todo #3: coin-selection, tests, retrying and errors
+
+### Comon backend features
+
+There is a list of common tasks shared by multiple backends,
+which could be tackled generically in our framework.
+
+* Parsing transaction back to state-machine transition
+is required for delegated architecures,
```

Comment:
What are delegated architectures? Have them got to do with Hydra delegation topology.

Reply:
Delegated architecture is one where each client only contacts chain, in that case L2, thought Delegate servers. That is simplest and only architecture working right now for Hydra. We spawn bunch of Hydra Nodes, each containing Delegate server accepting client requests. I am saying about old Hydra Auction or probably any Hydra baked project for now.

This Delegate server should keep track of Hydra state overall and Tx pushed on it specifically. Which gives a need to parse Tx from intended DApp protocol, to keep its current state. For example: Auction has CurrentBid state, which delegate server should be aware of. This means some DApp Tx-parsing should be present. We aim to give generic utils for this.